python/cudf/cudf/core/udf/pipeline.py

```diff
@@ -50,3 +66,168 @@ def wrapper(*args):
+    actually JIT the kernel itself later, accounting for types
+    and offsets
+    """
+
+    # Tuple of arrays, first the output data array, then the mask
+    return_type = Tuple((return_type[::1], boolean[::1]))
+    offsets = []
+    sig = [return_type]
+    for col in df._data.values():
+        sig.append(masked_arrty_from_np_type(col.dtype))
+        offsets.append(int64)
+
+    # return_type + data,masks + offsets + size
+    sig = void(*(sig + offsets + [int64]))
+
+    return sig
+
+
+@cuda.jit(device=True)
+def mask_get(mask, pos):
+    return (mask[pos // mask_bitsize] >> (pos % mask_bitsize)) & 1
+
+
+kernel_template = """\
+def _kernel(retval, {input_columns}, {input_offsets}, size):
+    i = cuda.grid(1)
+    ret_data_arr, ret_mask_arr = retval
+    if i < size:
+{masked_input_initializers}
+        ret = {user_udf_call}
+        ret_data_arr[i] = {ret_value}
+        ret_mask_arr[i] = {ret_valid}
+"""
+
+unmasked_input_initializer_template = """\
+        d_{idx}, m_{idx} = input_col_{idx}
+        masked_{idx} = Masked(d_{idx}[i], True)
+"""
+
+masked_input_initializer_template = """\
+        d_{idx}, m_{idx} = input_col_{idx}
+        masked_{idx} = Masked(d_{idx}[i], mask_get(m_{idx}, i + offset_{idx}))
+"""
+
+
+@annotate("DEFINE", color="yellow", domain="cudf_python")
+def _define_function(df, scalar_return=False):
+    # Create argument list for kernel
+    input_columns = ", ".join(
+        [f"input_col_{i}" for i in range(len(df.columns))]
+    )
+
+    input_offsets = ", ".join([f"offset_{i}" for i in range(len(df.columns))])
+
+    # Create argument list to pass to device function
+    args = ", ".join([f"masked_{i}" for i in range(len(df.columns))])
+    user_udf_call = f"f_({args})"
+
+    # Generate the initializers for each device function argument
+    initializers = []
+    for i, col in enumerate(df._data.values()):
+        idx = str(i)
+        if col.mask is not None:
+            template = masked_input_initializer_template
+        else:
+            template = unmasked_input_initializer_template
+
+        initializer = template.format(**{"idx": idx})
+
+        initializers.append(initializer)
+
+    masked_input_initializers = "\n".join(initializers)
+
+    # Generate the code to extract the return value and mask from the device
+    # function's return value depending on whether it's already masked
+    if scalar_return:
+        ret_value = "ret"
+        ret_valid = "True"
+    else:
+        ret_value = "ret.value"
+        ret_valid = "ret.valid"
+
+    # Incorporate all of the above into the kernel code template
+    d = {
+        "input_columns": input_columns,
+        "input_offsets": input_offsets,
+        "masked_input_initializers": masked_input_initializers,
+        "user_udf_call": user_udf_call,
+        "ret_value": ret_value,
+        "ret_valid": ret_valid,
+    }
+
+    return kernel_template.format(**d)
+
+
+@annotate("UDF PIPELINE", color="black", domain="cudf_python")
```

Comment:
```suggestion
@annotate("UDF PIPELINE", color="black", domain="cudf_python")
```

`black` is unfortunately not a color we support (unless matplotlib is installed). I believe it would also be difficult (impossible) to read in the profiles.

https://github.com/NVIDIA/NVTX/blob/dev/python/nvtx/colors.py#L9